File: lib/data/models/student.dart
class Student {
  final String id;
  final String name;
  final String nis;
  final String className;
  final String gender;
  final String? photoUrl;
  int totalViolationPoints;
  int totalAchievementPoints;
  int violationCount;
  int achievementCount;

  Student({
    required this.id,
    required this.name,
    required this.nis,
    required this.className,
    required this.gender,
    this.photoUrl,
    this.totalViolationPoints = 0,
    this.totalAchievementPoints = 0,
    this.violationCount = 0,
    this.achievementCount = 0,
  });

  factory Student.fromJson(Map<String, dynamic> json) {
    return Student(
      id: json['id'].toString(),
      name: json['name'] ?? '',
      nis: json['nis'] ?? '',
      className: json['class_name'] ?? '',
      gender: json['gender'] ?? 'L',
      photoUrl: json['photo_url'],
      totalViolationPoints: json['total_violation_points'] ?? 0,
      totalAchievementPoints: json['total_achievement_points'] ?? 0,
      violationCount: json['violation_count'] ?? 0,
      achievementCount: json['achievement_count'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'nis': nis,
      'class_name': className,
      'gender': gender,
      'photo_url': photoUrl,
    };
  }

  int get behaviorScore => 100 - totalViolationPoints + totalAchievementPoints;

  String get behaviorGrade {
    final score = behaviorScore;
    if (score >= 90) return 'A';
    if (score >= 80) return 'B';
    if (score >= 70) return 'C';
    if (score >= 60) return 'D';
    return 'E';
  }

  String get initials {
    final parts = name.split(' ');
    if (parts.length >= 2) {
      return '${parts[0][0]}${parts[1][0]}'.toUpperCase();
    }
    return name.substring(0, 2).toUpperCase();
  }
}

File: backend/app/Models/Student.php
<?php

namespace App\Models;

use Illuminate\Foundation\Auth\User as Authenticatable;
use Laravel\Sanctum\HasApiTokens;
use Illuminate\Database\Eloquent\Relations\HasMany;

class Student extends Authenticatable
{
    use HasApiTokens;

    protected $fillable = [
        'name',
        'nis',
        'class_name',
        'gender',
        'photo_url',
        'password',
        'total_violation_points',
        'total_achievement_points',
        'violation_count',
        'achievement_count',
    ];

    protected $hidden = [
        'password',
    ];

    protected $casts = [
        'total_violation_points' => 'integer',
        'total_achievement_points' => 'integer',
        'violation_count' => 'integer',
        'achievement_count' => 'integer',
    ];

    public function violations(): HasMany
    {
        return $this->hasMany(Violation::class);
    }

    public function achievements(): HasMany
    {
        return $this->hasMany(Achievement::class);
    }

    public function getBehaviorScoreAttribute(): int
    {
        return 100 - $this->total_violation_points + $this->total_achievement_points;
    }

    public function getBehaviorGradeAttribute(): string
    {
        $score = $this->behavior_score;
        if ($score >= 90) return 'A';
        if ($score >= 80) return 'B';
        if ($score >= 70) return 'C';
        if ($score >= 60) return 'D';
        return 'E';
    }

    public function recalculatePoints(): void
    {
        $this->total_violation_points = $this->violations()->sum('points');
        $this->violation_count = $this->violations()->count();
        $this->total_achievement_points = $this->achievements()->sum('points');
        $this->achievement_count = $this->achievements()->count();
        $this->save();
    }
}
